addons/spectator/functions/fnc_cam_draw3D.sqf

```diff
@@ -0,0 +1,193 @@
+GVAR(camDeltaTime) = _currentTime - GVAR(camLastTickTime);
+GVAR(camLastTickTime) = _currentTime;
+
+// If no focus in unit camera modes try to find a new one
+if (_cameraMode in MODES_UNIT) then {
+    private _focus = if (isNull _camTarget) then {
+        private _testFocus = ([] call FUNC(getTargetEntities)) select 0;
+        if (isNil "_testFocus") then {
+            objNull
+        } else {
+            _testFocus
+        }
+    } else {
+        _camTarget
+    };
+
+    // If new focus was found then switch to it
+    if !(isNull _focus && {_focus != _camTarget}) then {
+        [_focus] call FUNC(setFocus);
+    };
+
+    if (!isNull _focus && {_cameraMode == MODE_FOLLOW}) then {
+        [_focus] call FUNC(cam_prepareTarget);
+    };
+};
+
+// Refresh the local variable
+_camTarget = GVAR(camTarget);
+
+// Focus get in / out of vehicle state
+if !(isNull _camTarget) then {
+    private _targetInVeh = GVAR(camTargetInVehicle);
+
+    if (GVAR(camHasTarget)) then {
+        if (!_targetInVeh && { vehicle _camTarget != _camTarget }) then {
+            [_camTarget] call FUNC(cam_setTarget);
+            GVAR(camTargetInVehicle) = true;
+        };
+
+        if (_targetInVeh && { vehicle _camTarget == _camTarget }) then {
+            [_camTarget] call FUNC(cam_setTarget);
+            GVAR(camTargetInVehicle) = false;
+        };
+    };
+} else {
+    GVAR(camTargetInVehicle) = false;
+};
+
+// Camera lights
+if (count GVAR(camLights) > 1) then {
+    (GVAR(camLights) select 1) setPosASL (AGLToASL (screenToWorld getMousePosition));
+};
+END_COUNTER(camTick);
+
+if !(GVAR(uiMapVisible)) then {
+    if (GVAR(drawUnits)) then {
+        BEGIN_COUNTER(drawTags);
+        // Groups and Units
+        {
+            _x params ["_unit", "_type", "_icon"];
+            private _position = (_unit modelToWorldVisual (_unit selectionPosition "Head")) vectorAdd [0,0,HEIGHT_OFFSET];
+
+            if (_type == 2 && { _unit distance _camera < DISTANCE_NAMES } && {_unit in _camTarget || _unit in _cursorObject}) then {
+                drawIcon3D [
+                    ICON_BACKGROUND_UNIT,
+                    [0, 0, 0, [0.4, 0.8] select (_unit in _camTarget)],
+                    _position,
+                    5.0,
+                    4,
+                    0.0,
+                    "",
+                    0,
+                    0.035,
+                    "PuristaMedium",
+                    "center"
+                ];
+            };
+
+            // Apply modifiers
+            if (_type == 1 && { time <= GETVAR(_unit,GVAR(highlightTime),0) }) then {
+                _icon set [1, [1,1,1, ((_icon select 1) select 3)]];
+            };
+            _icon set [2, _position];
+
+            // Draw icon
+            drawIcon3D _icon;
+
+            nil // Speed loop
+        } count GVAR(thingsToDraw);
+        END_COUNTER(drawTags);
+    };
+
+    // Draw projectiles and grenades paths
+    if (GVAR(drawProjectiles)) then {
```

Comment:
Makes sense to me.

Though, you might still [add the unit](https://github.com/CBATeam/CBA_A3/blob/master/addons/diagnostic/fnc_addUnitTrackProjectiles.sqf) when in proximity and [remove it](https://github.com/CBATeam/CBA_A3/blob/master/addons/diagnostic/fnc_removeUnitTrackProjectiles.sqf) when too far from it.
And for performance you could [reduce the amount of lines that are drawn at the same time](https://github.com/CBATeam/CBA_A3/blob/master/addons/diagnostic/XEH_preInit.sqf#L15)